enumerables.rb

```diff
@@ -0,0 +1,243 @@
+# rubocop:disable Metrics/ModuleLength
+# rubocop:disable Metrics/CyclomaticComplexity
+# rubocop:disable Metrics/PerceivedComplexity
+module Enumerable
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    index = 0
+    h_array = to_a
+
+    if instance_of?(Range) == true
+      step do
+        yield(h_array[index])
+        index += 1
+      end
+    end
+    return self unless instance_of?(Range) == false
+
+    0.upto(length - 1) do |v|
+      yield(h_array[v])
+    end
+    self
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    h_array = to_a
+    0.upto(length - 1) do |v|
+      yield(h_array[v], v)
+    end
+  end
+
+  def my_select
+    return to_enum(:my_select) unless block_given?
+
+    h_array = to_a
+    new_arr = []
+    var = 0
+    h_array.my_each do |v|
+      if yield v
+        new_arr[var] = v
+        var += 1
+      end
+    end
+    new_arr
+  end
+
+  def my_all?(num = nil)
+    h_array = to_a
+
+    h_array.my_each do |v|
+      break unless num.nil? == false
+      return false unless (v == num) == true
+    end
+    return true unless block_given?
+
+    unless block_given?
+      h_array.my_each do |v|
+        return false unless v == true
+      end
+    end
+    return true unless block_given?
+
+    h_array.my_each do |v|
+      return false unless yield(v)
+    end
+    true
+  end
+
+  def my_any?
+    h_array = to_a
+    h_array.my_each do
+      return true unless yield(v) == false
+    end
+    false
+  end
+
+  def my_none?(num = nil)
+    h_array = to_a
+
+    h_array.my_each do |v|
+      break unless num.nil? == false
+      return false unless (v == num) == false
+    end
+    return true unless block_given?
+
+    h_array.my_each do |v|
+      return false unless yield(v) == false
+    end
+    true
+  end
+
+  def my_count(num = '')
+    h_array = to_a
+    count = 0
+    h_array.my_each do
+      count += 1
+    end
+    return count unless (num == '') == false || block_given?
+
+    num.to_i
+    count = 0
+    h_array.my_each do |v|
+      count += 1 unless (v == num) == false
+    end
+    return count unless block_given?
+
+    count = 0
+    h_array.my_each do |v|
+      count += 1 unless yield(v) == false
+    end
+    count
+  end
+
+  def my_map
+    return to_enum(:double) unless block_given?
+
+    h_array = to_a
+    map_arr = []
+
+    h_array.my_each do |v|
+      map_arr.push(yield(v))
+    end
+    map_arr
+  end
```

Comment:
- **_`#my_map` should execute only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum._**

Reply:
We checked this also and the original #map function does not prioritize a proc over a block. The original #map function returns an error if you pass it a proc and a block at the same time. And so does our #my_map function.